### PR TITLE
Np 2157 use commons

### PR DESCRIPTION
**Pull request by torarnet**

### CODE REVIEW

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java [hunk 1]
```diff
@@ -45,75 +36,70 @@
...
         this.cristinApiClient = cristinApiClient;
     }
 
-    public void setCristinApiClient(CristinApiClient cristinApiClient) {
-        this.cristinApiClient = cristinApiClient;
+    @Override
+    protected ProjectPresentation[] processInput(Void input, RequestInfo requestInfo, Context context)
+        throws ApiGatewayException {
+
+        return createProjectPresentations(checkLanguageAndFetch(requestInfo), checkTitleAndFetch(requestInfo));
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:59]: assign the parameters to variables for better readbility.

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java [hunk 2]
```diff
@@ -45,75 +36,70 @@
...
+        List<Project> projects = attempt(() ->
+            cristinApiClient.queryAndEnrichProjects(cristinQueryParameters, language)).orElseThrow();
 
-        String language = queryStringParameters.getOrDefault(LANGUAGE_KEY, DEFAULT_LANGUAGE_CODE);
-        if (!VALID_LANGUAGE_CODES.contains(language)) {
-            throw new RuntimeException(LANGUAGE_INVALID);
-        }
+        return projects.stream()
+            .map(project -> presentationConverter.asProjectPresentation(project, language))
+            .collect(Collectors.toList());
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:102]: refactor to method `transfromProjectsToProjectPresentations` ?

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java [hunk 3]
```diff
@@ -128,10 +114,14 @@ private boolean isValidTitle(String str) {
 
     private boolean isValidCharacter(char c) {
         return Character.isWhitespace(c)
-                || Character.isLetterOrDigit(c)
-                || c == CHARACTER_DASH
-                || c == CHARACTER_COMMA
-                || c == CHARACTER_PERIOD;
+            || Character.isLetterOrDigit(c)
+            || c == CHARACTER_DASH
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:118]: These 3 chars could be a set of invalid characters and then you only  need to check INVALID_CHARS.contains(c)

**Review by axthosarouris — APPROVED** [review]

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java (hunk 3 above)
Comment (torarnet) [reply src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:118]: Those 3 chars are actually valid characters in addition to letter, digit and whitespace. But i can make a set of those character and do a VALID_CHARS.contains(c)

File: build.gradle [hunk 4]
```diff
@@ -13,11 +13,30 @@ targetCompatibility = JavaVersion.VERSION_11 // bytecode target version
 
 repositories {
     mavenCentral()
+    maven { url  "https://dl.bintray.com/unit-nva/nva-commons" }
+    maven { url 'https://jitpack.io' }
+}
+
+project.ext{
+    nvaCommonsVersion = '1.0.4'
```
Comment (brinxmat) [inline build.gradle:21]: It surprises me that jupiterVersion is not in project.ext, but this is not to say that it is incorrect

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java [hunk 5]
```diff
@@ -45,75 +36,70 @@
...
+    private String checkLanguageAndFetch(RequestInfo requestInfo) throws BadRequestException {
+        return Optional.of(getQueryParamOpt(requestInfo, LANGUAGE_QUERY_PARAMETER)
+            .orElse(DEFAULT_LANGUAGE_CODE))
+            .filter(this::isValidLanguage)
+            .orElseThrow(() -> new BadRequestException(LANGUAGE_INVALID));
+    }
 
-            Type projectListType = new TypeToken<ArrayList<ProjectPresentation>>() {
-            }.getType();
+    protected Optional<String> getQueryParamOpt(RequestInfo requestInfo, String queryParameter) {
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:80]: getQueryParameter, the return type is clear from the method definition

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java [hunk 6]
```diff
@@ -45,75 +36,70 @@
...
-        Map<String, String> queryStringParameters = (Map<String, String>) input.get(QUERY_STRING_PARAMETERS_KEY);
-        String title = queryStringParameters.get(TITLE_KEY);
-        String language = queryStringParameters.getOrDefault(LANGUAGE_KEY, DEFAULT_LANGUAGE_CODE);
+    protected Integer getSuccessStatusCode(Void input, ProjectPresentation[] output) {
+        return HttpURLConnection.HTTP_OK;
+    }
 
-        try {
-            Map<String, String> cristinQueryParameters = createCristinQueryParameters(title, language);
+    private String checkTitleAndFetch(RequestInfo requestInfo) throws BadRequestException {
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:67]: getValidTitle (I don't like methods that tell me they do multiple things with "and" as this tends to mean the method is violating the SRP), I also tend to reserve "fetch" for HTTP-GETs.

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java [hunk 7]
```diff
@@ -45,75 +36,70 @@
...
-            gatewayResponse.setStatusCode(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode());
-            gatewayResponse.setErrorBody(e.getMessage());
-        }
+    private ProjectPresentation[] createProjectPresentations(String language, String title) {
+        List<ProjectPresentation> projectPresentations = createProjectPresentationList(language, title);
+        return convertToArray(projectPresentations);
+    }
 
-        return gatewayResponse;
+    private ProjectPresentation[] convertToArray(List<ProjectPresentation> projectPresentations) {
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:89]: Probably just
        return projectPresentations.toArray(ProjectPresentation[]::new);

**Review by brinxmat — CHANGES_REQUESTED** [review]
In general, very tidy, well formulated code. Follows our standards

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java (hunk 3 above)
Comment (torarnet) [reply src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:118]: Seems char array works a little different, so after discussing it with Orestis we decided to leave it as it is

File: build.gradle [hunk 8]
```diff
@@ -13,11 +13,30 @@ targetCompatibility = JavaVersion.VERSION_11 // bytecode target version
...
+    jacksonVersion = '2.12.0'
 }
 
 def jupiterVersion = "5.6.0"
 
 dependencies {
+    testImplementation 'com.github.BIBSYSDEV:nva-testutils:0.1.18'
```
[the remark this answers is not in the document]
Comment (torarnet) [reply build.gradle:28]: sort them last in each group or put every testDependency last under all other dependencies?

File: build.gradle (hunk 4 above)
Comment (torarnet) [reply build.gradle:21]: Should i put it there?

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java (hunk 6 above)
Comment (torarnet) [reply src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:67]: should i only change name or split logic into one validator and one get value as well? I wanted to do several things with less code, and calling this::isValid implicit made the code a little more compact

File: build.gradle (hunk 4 above)
Comment (brinxmat) [reply build.gradle:21]: Why not? :)

**Review by brinxmat — APPROVED** [review]

File: build.gradle (hunk 4 above)
Comment (torarnet) [reply build.gradle:21]: Moved it to project.ext

**Review by axthosarouris — APPROVED** [review]

**Pull request merged**